Allow tests to share a single Spark context.

tomwhite opened the pull request:
It's not possible to have multiple Spark contexts per JVM (even if not running concurrently),
and while one solution is to fork a new JVM per test class, it's not always possible or desirable
to do this - if for example the test suite has a dataflow test per test method, or the test
suite has many very small tests.

This JIRA has more background and discussion about the JVM limitation:

https://issues.apache.org/jira/browse/SPARK-2243

This change makes it possible to share a SparkContext when running tests that use the
SparkPipelineRunner by setting the system property 'dataflow.spark.test.reuseSparkContext'
to 'true'.

srowen commented:
In the past, rather than do this, I've just used a separate JVM for each test suite. That way they can run in parallel even. The price is some overhead, but gains some isolation and no need to try to manage the context like this.

tomwhite commented:
Ideally we would use a separate VM per test suite, but we need a way of being able to reliably close the context per test method, and unfortunately dataflow doesn't offer that since the Pipeline class has no `close()` method. We have one on the return value from the run() method, but other implementations of dataflow do not, so we still need something like this PR to allow other tests suites to run with spark-dataflow. (I have opened an issue against Cloud Dataflow here: https://github.com/GoogleCloudPlatform/DataflowJavaSDK/issues/42)

I've changed spark-dataflow's test suite back to use a separate JVM per suite though. I've also added calls to close the context for tests that didn't have that.

Thanks for your feedback @srowen - it's very helpful.
